corehq/form_processor/models/forms.py

```diff
@@ -423,6 +406,32 @@ def hard_delete_forms(self, domain, form_ids, delete_attachments=True, *, publis
 
         return deleted_count
 
+    def hard_delete_forms_before_cutoff(self, cutoff, dry_run=True):
+        """
+        Permanently deletes forms with deleted_on set to a datetime earlier than
+        the specified cutoff datetime and creates a tombstone record of the deletion.
+        :param cutoff: datetime used to obtain the forms to be hard deleted
+        :param dry_run: if True, no changes will be committed to the database
+        and this method is effectively read-only
+        :return: dictionary of count of deleted objects per table
+        """
+        counts = {}
+        class_path = 'form_processor.XFormInstance'
+        for db_name in get_db_aliases_for_partitioned_query():
+            queryset = self.using(db_name).filter(deleted_on__lt=cutoff)
+            if dry_run:
+                deleted_counts = {class_path: queryset.count()}
+                counts.update(deleted_counts)
+            else:
+                form_tombstones = [DeletedSQLDoc(doc_id=form.form_id, object_class_path=class_path,
+                                                 domain=form.domain, deleted_on=form.deleted_on)
+                                   for form in queryset]
+                for chunk in chunked(form_tombstones, 1000, list):
+                    DeletedSQLDoc.objects.bulk_create(chunk, ignore_conflicts=True)
+                deleted_counts = queryset.delete()[1]
```

Comment:
Similar to how you override the CommCareCase.delete method to ensure tombstones are created, you could do the same thing on the CommCareCaseManager's delete method for bulk deletes. What do you think?

Reply:
ohhh cool I thought you couldn't override the delete function for queryset.delete but I guess you can. The one hesitation I have about this is that I'd have exempt all existing processes for domain deletion and for deleting system cases/forms (like the location cases) from creating tombstones. Though maybe something as simple as passing `is_domain_or_system_deletion=True` would be enough (after I find them all).